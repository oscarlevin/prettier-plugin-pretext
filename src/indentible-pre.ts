/**
 * Return the amount of leading whitespace in `str`.
 */
function getLeadingWhitespace(str: string, tabWidth = 4): number {
    if (str.trim() === "") {
        return Number.MAX_VALUE;
    }
    let whitespace = 0;
    for (const char of str) {
        if (char.match(" ")) {
            whitespace += 1;
            continue;
        }
        if (char.match("\t")) {
            whitespace += tabWidth;
            continue;
        }
        break;
    }

    return whitespace;
}

/**
 * Get the number of characters which need to be trimmed off to trim off `maxWhitespaceLen` amount
 * of whitespace. This function takes care to count tabs as multiple whitespaces but as only one character.
 */
function getNumCharsForWhitespaceLen(
    str: string,
    maxWhitespaceLen: number,
    tabWidth = 4
): number {
    let chars = 0;
    let whitespace = 0;
    for (const c of str) {
        if (c.match(" ")) {
            chars += 1;
            whitespace += 1;
        } else if (c.match("\t")) {
            chars += 1;
            whitespace += tabWidth;
        } else {
            break;
        }

        if (whitespace > maxWhitespaceLen) {
            return chars - 1;
        }
        if (whitespace === maxWhitespaceLen) {
            return chars;
        }
    }
    return chars;
}

/**
 * Strip a uniform amount of space off the front of each line and return the lines separated
 * in an array.
 */
export function indentablePreToLines(rawText: string, tabWidth = 4): string[] {
    let lines = rawText.split("\n");
    const charsTrimmable = Math.min(
        ...lines.map((l) => getLeadingWhitespace(l, tabWidth))
    );
    if (charsTrimmable === 0) {
        return lines;
    }

    lines = lines.map((line) => {
        const trimPos = getNumCharsForWhitespaceLen(
            line,
            charsTrimmable,
            tabWidth
        );
        return line.slice(trimPos);
    });

    // We trim off any leading and trailing whitespace-only lines
    let leadingBlankLines = 0;
    let trailingBlankLines = 0;
    for (let i = 0; i < lines.length; i++) {
        if (lines[i].trim() === "") {
            leadingBlankLines += 1;
        } else {
            break;
        }
    }
    for (let i = lines.length - 1; i > 0; i--) {
        if (lines[i].trim() === "") {
            trailingBlankLines += 1;
        } else {
            break;
        }
    }

    return lines.slice(leadingBlankLines, lines.length - trailingBlankLines);
}
