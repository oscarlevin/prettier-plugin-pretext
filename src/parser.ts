import { parse as xmlToolsParse } from "@xml-tools/parser";
import type { Parser } from "./types";

type LocatedError = Error & {
    loc: {
        start: { line: number; column: number };
        end: { line: number; column: number };
    };
};

const parser: Parser = {
    parse(text) {
        const { lexErrors, parseErrors, cst } = xmlToolsParse(text);

        // If there are any lexical errors, throw the first of them as an error.
        if (lexErrors.length > 0) {
            const lexError = lexErrors[0];
            const error = new Error(lexError.message) as LocatedError;

            error.loc = {
                start: { line: lexError.line, column: lexError.column },
                end: {
                    line: lexError.line,
                    column: lexError.column + lexError.length,
                },
            };

            throw error;
        }

        // If there are any parse errors, throw the first of them as an error.
        if (parseErrors.length > 0) {
            const parseError = parseErrors[0];
            const error = new Error(parseError.message) as LocatedError;

            const { token } = parseError;
            error.loc = {
                start: { line: token.startLine!, column: token.startColumn! },
                end: { line: token.endLine!, column: token.endColumn! },
            };

            throw error;
        }

        // Otherwise return the CST.
        return cst;
    },
    astFormat: "ptx",
    locStart(node) {
        return node.location!.startOffset;
    },
    locEnd(node) {
        return node.location!.endOffset!;
    },
};

export default parser;
