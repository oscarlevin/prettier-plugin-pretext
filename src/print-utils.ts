import { Fragment, IToken, Path } from "./types";

export function printIToken(path: Path<IToken>): Fragment {
    const node = path.getValue();

    return {
        offset: node.startOffset,
        startLine: node.startLine,
        endLine: node.endLine,
        printed: node.image,
        endOffset: node.endOffset,
    };
}

export function isFragment(item: any): item is Fragment {
    if (typeof item === "object" && item.offset != null) {
        return true;
    }
    return false;
}
