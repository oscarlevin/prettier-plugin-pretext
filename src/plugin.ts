import type { Plugin } from "./types";
import parser from "./parser";
import printer from "./printer";

const plugin: Plugin = {
    languages: [
        {
            name: "PTX",
            parsers: ["ptx"],
            extensions: [".xml", ".ptx"],
            vscodeLanguageIds: ["xml", "ptx"],
            linguistLanguageId: 399,
        },
    ],
    parsers: {
        ptx: parser,
    },
    printers: {
        ptx: printer,
    },
    options: {},
    defaultOptions: {
        printWidth: 80,
        tabWidth: 4,
        useTabs: true,
    },
};

export = plugin;
