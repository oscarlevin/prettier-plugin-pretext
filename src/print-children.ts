import { builders } from "prettier/doc";
import { printIToken } from "./print-utils";
import type { AnyNode, Fragment, Path, Print } from "./types";

const { fill, group, hardline, indent, join, line, literalline, softline } =
    builders;

/**
 * Prints the children of `path` but trims whitespace from the start/end and collapses
 * contiguous whitespace into a single char.  Inserts `line` between words, and `hardline`
 * after sentence-ending punctuation (.!?).
 */
export function printChildrenWhenWhitespaceDoesNotMatter(
    path: Path<AnyNode>,
    print: Print
) {
    const node = path.getNode();
    if (node == null || node.name !== "element") {
        throw new Error("Expected node with name `element`");
    }
    const nodePath = path as Path<typeof node>;

    const fragments = nodePath.call(
        (childrenPath) => {
            const children = childrenPath.getValue();
            const response: Fragment[] = [];

            if (children.Comment) {
                response.push(...childrenPath.map(printIToken, "Comment"));
            }

            if (children.chardata) {
                childrenPath.each((charDataPath) => {
                    const chardata = charDataPath.getValue();
                    if (!chardata.children.TEXT) {
                        return;
                    }

                    const rawText = chardata.children.TEXT[0].image;
                    const leadingWhitespace = rawText.match(/^\s*/)![0].length;
                    const trailingWhitespace = rawText.match(/\s*$/)![0].length;
                    const content = rawText.trim();
                    const printed = fill(
                        content.split(/(\s+|[.!?]\s+)/g).map((value) => {
                            if (value.match(/[.!?]\s+/)) {
                                // This will put a linebreak after any sentence-ending punctuation in chardata.
                                return [value.trim(), hardline];
                            } else if (value.match(/\s+/)) {
                                return line;
                            }
                            return value;
                        })
                    );

                    const location = chardata.location!;
                    response.push({
                        offset: location.startOffset + leadingWhitespace,
                        startLine: location.startLine,
                        endLine: location.endLine!,
                        endOffset: location.endOffset! - trailingWhitespace,
                        nodeType: "text",
                        printed,
                    } as Fragment);
                }, "chardata");
            }

            if (children.element) {
                response.push(
                    ...childrenPath.map((elementPath) => {
                        const node = elementPath.getValue();
                        const location = node.location!;
                        const tagName = node.children.Name[0]?.image?.trim();

                        return {
                            offset: location.startOffset,
                            startLine: location.startLine,
                            endLine: location.endLine!,
                            endOffset: location.endOffset,
                            nodeType: "element",
                            tagName,
                            printed: print(elementPath),
                        } as Fragment;
                    }, "element")
                );
            }

            if (children.PROCESSING_INSTRUCTION) {
                response.push(
                    ...childrenPath.map(printIToken, "PROCESSING_INSTRUCTION")
                );
            }

            return response;
        },
        "children",
        "content",
        0,
        "children"
    );

    fragments.sort((left, right) => left.offset - right.offset);

    return fragments;
}
