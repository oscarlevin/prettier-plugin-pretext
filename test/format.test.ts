import fs from "fs";
import path from "path";
import prettier from "prettier";

import { Options } from "../src/types";
import plugin from "../src/plugin";

const fixture = fs.readFileSync(path.join(__dirname, "./fixture.ptx"), "utf-8");
const example2 = fs.readFileSync(
    path.join(__dirname, "./example2.ptx"),
    "utf-8"
);

function format(content: string, opts: Partial<Options> = {}) {
    return prettier.format(content, {
        ...opts,
        parser: "ptx",
        plugins: [plugin as any as string], // hacky but it works
    });
}

test("defaults", () => {
    const formatted = format(fixture);
    expect(formatted).toMatchSnapshot();
});

test("basic pretext source", () => {
    const formatted = format(example2);
    expect(formatted).toMatchSnapshot();
});


test("bracketSameLine => true", () => {
    const formatted = format(fixture, {
        bracketSameLine: true,
    });

    expect(formatted).toMatchSnapshot();
});

test("xmlSelfClosingSpace => false", () => {
    const formatted = format(fixture, {
        xmlSelfClosingSpace: false,
    });

    expect(formatted).toMatchSnapshot();
});

test("bracketSameLine => true, xmlSelfClosingSpace => false", () => {
    const formatted = format(fixture, {
        bracketSameLine: true,
        xmlSelfClosingSpace: false,
    });

    expect(formatted).toMatchSnapshot();
});

test("singleAttributePerLine => true", () => {
    const formatted = format(fixture, {
        singleAttributePerLine: true,
    });

    expect(formatted).toMatchSnapshot();
});
