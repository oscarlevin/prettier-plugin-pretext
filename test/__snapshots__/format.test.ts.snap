// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic pretext source 1`] = `
"<section>
	<p>A video is a <br/> natural way to</p>
	<p>
		A video is a natural <br/> way to enhance a document when rendered in an
		electronic format, such as HTML web pages. It might be additional
		information that is hard to communicate with text (marine invertebrates
		swimming), a
	</p>
	<p>
		Some <trademark/> text <m>math and stuff</m>. and
		<ul>
			<li>foo</li>
			<li>bar</li>
		</ul>
		and more textsuch as HTML web pages. It might be additional information
		that is hard to communicate with text (marine invertebrates swimming), a
	</p>
	<sage>
		<output>fjdklsafjlkdjlkflk  fdjal f fjdlasjlf
     jfdajldlsldk fjlkfajslfjllkadkf                ba</output>
		<input>
                A = matrix(4,5, srange(20)) <br />a really really really long line i think that it is long<br />
                 A.rref()
   </input>
	</sage>
	<p>
		fun times
		<ul>
			<li>foo</li>
			<li>bar</li>
		</ul>
		with list
	</p>
	<p>
		fun times
		<ul>
			<li>foo</li>
			<li>bar</li>
		</ul>

		with list
	</p>
</section>
"
`;

exports[`bracketSameLine => true 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE module PUBLIC \\"-//Puppy Crawl//DTD Check Configuration 1.3//EN\\"
	\\"https://www.puppycrawl.com/dtds/configuration_1_3.dtd\\">
<?xml-model href=\\"project.rnc\\" type=\\"application/relax-ng-compact-syntax\\"?>
<!-- foo -->
<svg
	xmlns=\\"http://www.w3.org/2000/svg\\"
	xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"
	width=\\"200\\"
	height=\\"100\\"
	viewBox=\\"0 0 200 100\\"
>
	<title>Style inheritance and the use element</title>
	<desc _attr=\\"attr\\">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
	<?pagebreak?>

	<style/>
	<style/>
	<style type=\\"text/css\\">
circle {
	stroke-opacity: 0.7;
}
.special circle {
	stroke: green;
}
use {
	stroke: purple;
	fill: orange;
}
	</style>

	<yaml
		myveryveryveryverylongattributename=\\"myveryveryveryverylongattributevalue\\"
	>
		- 1 - 2 - 3
	</style>

	<!-- inner comment -->

	<?pagebreak?>
	<g class=\\"special\\" style=\\"fill: blue\\">
		<circle id=\\"c\\" cy=\\"50\\" cx=\\"50\\" r=\\"40\\" stroke-width=\\"20\\"/>
	</g>
	<use xlink:href=\\"#c\\" x=\\"100\\"/>
	<ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
	<p>
		Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed at est eget
		enim consectetur accumsan. Aliquam pretium sodales ipsum quis dignissim.
		Sed id sem vel diam luctus fringilla. Aliquam quis egestas magna.
		Curabitur molestie lorem et odio porta, et molestie libero laoreet.
		Morbi rhoncus sagittis cursus. Nullam vehicula pretium consequat.
		Praesent porta ante at posuere sollicitudin. Nullam commodo tempor arcu,
		at condimentum neque elementum ut.
	</p>
	<span>content</span>

	<div>
		even more
		<content/>
	</div>
</svg>
<!-- bar -->
"
`;

exports[`bracketSameLine => true, xmlSelfClosingSpace => false 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE module PUBLIC \\"-//Puppy Crawl//DTD Check Configuration 1.3//EN\\"
	\\"https://www.puppycrawl.com/dtds/configuration_1_3.dtd\\">
<?xml-model href=\\"project.rnc\\" type=\\"application/relax-ng-compact-syntax\\"?>
<!-- foo -->
<svg
	xmlns=\\"http://www.w3.org/2000/svg\\"
	xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"
	width=\\"200\\"
	height=\\"100\\"
	viewBox=\\"0 0 200 100\\"
>
	<title>Style inheritance and the use element</title>
	<desc _attr=\\"attr\\">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
	<?pagebreak?>

	<style/>
	<style/>
	<style type=\\"text/css\\">
circle {
	stroke-opacity: 0.7;
}
.special circle {
	stroke: green;
}
use {
	stroke: purple;
	fill: orange;
}
	</style>

	<yaml
		myveryveryveryverylongattributename=\\"myveryveryveryverylongattributevalue\\"
	>
		- 1 - 2 - 3
	</style>

	<!-- inner comment -->

	<?pagebreak?>
	<g class=\\"special\\" style=\\"fill: blue\\">
		<circle id=\\"c\\" cy=\\"50\\" cx=\\"50\\" r=\\"40\\" stroke-width=\\"20\\"/>
	</g>
	<use xlink:href=\\"#c\\" x=\\"100\\"/>
	<ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
	<p>
		Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed at est eget
		enim consectetur accumsan. Aliquam pretium sodales ipsum quis dignissim.
		Sed id sem vel diam luctus fringilla. Aliquam quis egestas magna.
		Curabitur molestie lorem et odio porta, et molestie libero laoreet.
		Morbi rhoncus sagittis cursus. Nullam vehicula pretium consequat.
		Praesent porta ante at posuere sollicitudin. Nullam commodo tempor arcu,
		at condimentum neque elementum ut.
	</p>
	<span>content</span>

	<div>
		even more
		<content/>
	</div>
</svg>
<!-- bar -->
"
`;

exports[`defaults 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE module PUBLIC \\"-//Puppy Crawl//DTD Check Configuration 1.3//EN\\"
	\\"https://www.puppycrawl.com/dtds/configuration_1_3.dtd\\">
<?xml-model href=\\"project.rnc\\" type=\\"application/relax-ng-compact-syntax\\"?>
<!-- foo -->
<svg
	xmlns=\\"http://www.w3.org/2000/svg\\"
	xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"
	width=\\"200\\"
	height=\\"100\\"
	viewBox=\\"0 0 200 100\\"
>
	<title>Style inheritance and the use element</title>
	<desc _attr=\\"attr\\">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
	<?pagebreak?>

	<style/>
	<style/>
	<style type=\\"text/css\\">
circle {
	stroke-opacity: 0.7;
}
.special circle {
	stroke: green;
}
use {
	stroke: purple;
	fill: orange;
}
	</style>

	<yaml
		myveryveryveryverylongattributename=\\"myveryveryveryverylongattributevalue\\"
	>
		- 1 - 2 - 3
	</style>

	<!-- inner comment -->

	<?pagebreak?>
	<g class=\\"special\\" style=\\"fill: blue\\">
		<circle id=\\"c\\" cy=\\"50\\" cx=\\"50\\" r=\\"40\\" stroke-width=\\"20\\"/>
	</g>
	<use xlink:href=\\"#c\\" x=\\"100\\"/>
	<ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
	<p>
		Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed at est eget
		enim consectetur accumsan. Aliquam pretium sodales ipsum quis dignissim.
		Sed id sem vel diam luctus fringilla. Aliquam quis egestas magna.
		Curabitur molestie lorem et odio porta, et molestie libero laoreet.
		Morbi rhoncus sagittis cursus. Nullam vehicula pretium consequat.
		Praesent porta ante at posuere sollicitudin. Nullam commodo tempor arcu,
		at condimentum neque elementum ut.
	</p>
	<span>content</span>

	<div>
		even more
		<content/>
	</div>
</svg>
<!-- bar -->
"
`;

exports[`singleAttributePerLine => true 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE module PUBLIC \\"-//Puppy Crawl//DTD Check Configuration 1.3//EN\\"
	\\"https://www.puppycrawl.com/dtds/configuration_1_3.dtd\\">
<?xml-model href=\\"project.rnc\\" type=\\"application/relax-ng-compact-syntax\\"?>
<!-- foo -->
<svg
	xmlns=\\"http://www.w3.org/2000/svg\\"
	xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"
	width=\\"200\\"
	height=\\"100\\"
	viewBox=\\"0 0 200 100\\"
>
	<title>Style inheritance and the use element</title>
	<desc _attr=\\"attr\\">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
	<?pagebreak?>

	<style/>
	<style/>
	<style type=\\"text/css\\">
circle {
	stroke-opacity: 0.7;
}
.special circle {
	stroke: green;
}
use {
	stroke: purple;
	fill: orange;
}
	</style>

	<yaml
		myveryveryveryverylongattributename=\\"myveryveryveryverylongattributevalue\\"
	>
		- 1 - 2 - 3
	</style>

	<!-- inner comment -->

	<?pagebreak?>
	<g
		class=\\"special\\"
		style=\\"fill: blue\\"
	>
		<circle
			id=\\"c\\"
			cy=\\"50\\"
			cx=\\"50\\"
			r=\\"40\\"
			stroke-width=\\"20\\"
		/>
	</g>
	<use
		xlink:href=\\"#c\\"
		x=\\"100\\"
	/>
	<ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
	<p>
		Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed at est eget
		enim consectetur accumsan. Aliquam pretium sodales ipsum quis dignissim.
		Sed id sem vel diam luctus fringilla. Aliquam quis egestas magna.
		Curabitur molestie lorem et odio porta, et molestie libero laoreet.
		Morbi rhoncus sagittis cursus. Nullam vehicula pretium consequat.
		Praesent porta ante at posuere sollicitudin. Nullam commodo tempor arcu,
		at condimentum neque elementum ut.
	</p>
	<span>content</span>

	<div>
		even more
		<content/>
	</div>
</svg>
<!-- bar -->
"
`;

exports[`xmlSelfClosingSpace => false 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE module PUBLIC \\"-//Puppy Crawl//DTD Check Configuration 1.3//EN\\"
	\\"https://www.puppycrawl.com/dtds/configuration_1_3.dtd\\">
<?xml-model href=\\"project.rnc\\" type=\\"application/relax-ng-compact-syntax\\"?>
<!-- foo -->
<svg
	xmlns=\\"http://www.w3.org/2000/svg\\"
	xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"
	width=\\"200\\"
	height=\\"100\\"
	viewBox=\\"0 0 200 100\\"
>
	<title>Style inheritance and the use element</title>
	<desc _attr=\\"attr\\">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
	<?pagebreak?>

	<style/>
	<style/>
	<style type=\\"text/css\\">
circle {
	stroke-opacity: 0.7;
}
.special circle {
	stroke: green;
}
use {
	stroke: purple;
	fill: orange;
}
	</style>

	<yaml
		myveryveryveryverylongattributename=\\"myveryveryveryverylongattributevalue\\"
	>
		- 1 - 2 - 3
	</style>

	<!-- inner comment -->

	<?pagebreak?>
	<g class=\\"special\\" style=\\"fill: blue\\">
		<circle id=\\"c\\" cy=\\"50\\" cx=\\"50\\" r=\\"40\\" stroke-width=\\"20\\"/>
	</g>
	<use xlink:href=\\"#c\\" x=\\"100\\"/>
	<ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
	<p>
		Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed at est eget
		enim consectetur accumsan. Aliquam pretium sodales ipsum quis dignissim.
		Sed id sem vel diam luctus fringilla. Aliquam quis egestas magna.
		Curabitur molestie lorem et odio porta, et molestie libero laoreet.
		Morbi rhoncus sagittis cursus. Nullam vehicula pretium consequat.
		Praesent porta ante at posuere sollicitudin. Nullam commodo tempor arcu,
		at condimentum neque elementum ut.
	</p>
	<span>content</span>

	<div>
		even more
		<content/>
	</div>
</svg>
<!-- bar -->
"
`;
